src/rgen/Microsoft.Macios.Generator/BindingSourceGeneratorGenerator.cs

```diff
@@ -19,6 +29,140 @@ public void Initialize (IncrementalGeneratorInitializationContext context)
+	{
+		var provider = context.SyntaxProvider
+			.CreateSyntaxProvider (
+				static (s, _) => s is T,
+				(ctx, _) => GetDeclarationForSourceGen<T> (ctx))
+			.Where (t => t.BindingAttributeFound)
+			.Select ((t, _) => t.Declaration);
+
+		context.RegisterSourceOutput (context.CompilationProvider.Combine (provider.Collect ()),
+			((ctx, t) => GenerateCode (ctx, t.Left, t.Right)));
+	}
+
+	/// <summary>
+	/// Generic method that can be used to filter/match a BaseTypeDeclarationSyntax with the BindingTypeAttribute.
+	/// Because our generator is focused only on Enum, Interface and Class declarations we can use a generic method
+	/// that will match the type + the presence of the attribute.
+	/// </summary>
+	/// <param name="context">Context used by the generator.</param>
+	/// <typeparam name="T">The BaseTypeDeclarationSyntax we are interested in.</typeparam>
+	/// <returns></returns>
+	static (T Declaration, bool BindingAttributeFound) GetDeclarationForSourceGen<T> (GeneratorSyntaxContext context)
+		where T : BaseTypeDeclarationSyntax
+	{
+		var classDeclarationSyntax = Unsafe.As<T> (context.Node);
+
+		// Go through all attributes of the class.
+		foreach (AttributeListSyntax attributeListSyntax in classDeclarationSyntax.AttributeLists)
+			foreach (AttributeSyntax attributeSyntax in attributeListSyntax.Attributes) {
+				if (context.SemanticModel.GetSymbolInfo (attributeSyntax).Symbol is not IMethodSymbol attributeSymbol)
+					continue; // if we can't get the symbol, ignore it
+
+				string attributeName = attributeSymbol.ContainingType.ToDisplayString ();
+
+				// Check the full name of the [Binding] attribute.
+				if (attributeName == AttributesNames.BindingAttribute)
+					return (classDeclarationSyntax, true);
+			}
+
+		return (classDeclarationSyntax, false);
+	}
+
+	/// <summary>
+	/// Collect the using statements from the class declaration root syntaxt tree and add them to the string builder
+	/// that will be used to generate the code. This way we ensure that we have all the namespaces needed by the
+	/// generated code.
+	/// </summary>
+	/// <param name="tree">Root syntax tree of the base type declaration.</param>
+	/// <param name="sb">String builder that will be used for the generated code.</param>
+	static void CollectUsingStatements (SyntaxTree tree, TabbedStringBuilder sb)
+	{
+		// collect all using from the syntax tree, add them to a hash to make sure that we don't have duplicates
+		// and add those usings that we do know we need for bindings.
+		var usingDirectives = tree.GetRoot ()
+			.DescendantNodes ()
+			.OfType<UsingDirectiveSyntax> ()
+			.Select (d => d.Name.ToString ()).ToArray ();
+		var usingDirectivesToKeep = new HashSet<string> (usingDirectives) {
+			// add the using statements that we know we need and print them to the sb
+		};
+		foreach (var ns in usingDirectivesToKeep) {
+			if (string.IsNullOrEmpty (ns))
+				continue;
+			sb.AppendFormatLine ("using {0};", ns);
```

Comment:
We're in the 21st century now:

```suggestion
			sb.AppendLine ($"using {ns};");
```